### PR TITLE
Yolov5/one shot

### PR DESCRIPTION
**Pull request by rahul-tuli**
The goal of this PR is to add one shot support for `YOLOv5` models.
To accomplish this we add a `--one-shot`, CLI argument, to `train.py` script. The functionality was tested with a locally build yolov5 wheel containing changes from [this PR](https://github.com/neuralmagic/yolov5/pull/46) 

Steps to reproduce test:
- [X] Checkout [this PR](https://github.com/neuralmagic/yolov5/pull/46) and cd into `yolov5`
- [X] Build the wheel `bash build_wheel.sh`
- [X] Now install the wheel in `sparseml venv`
- [X] Train a model in one-shot using the following example command
```bash
sparseml.object_detection.train \
    --weights " " --data coco128.yaml \
    --epochs 5 --cfg models_v5.0/yolov5s.yaml \
    --one-shot \
    --recipe zoo:cv/detection/yolov5-s/pytorch/ultralytics/coco/pruned-aggressive_96?recipe_type=transfer_learn
```

Relevant Output:
```bash
Applied recipe zoo:cv/detection/yolov5-s/pytorch/ultralytics/coco/pruned-aggressive_96?recipe_type=transfer_learn in One-Shot
Scaled weight_decay = 0.0005
optimizer: SGD with parameter groups 59 weight (no decay), 62 weight, 62 bias
train: Scanning '/XXXXXXXX/virtual_environments/sparseml3.8/lib/python3.8/site
val: Scanning '/XXXXXXXX/virtual_environments/sparseml3.8/lib/python3.8/site-p
Plotting labels to /XXXXXXXX/projects/sparseml/src/sparseml/pytorch/object_detection/runs/train/exp9/labels.jpg... 

AutoAnchor: 4.27 anchors/target, 0.994 Best Possible Recall (BPR). Current anchors are a good fit to dataset ✅
Image sizes 640 train, 640 val
Using 8 dataloader workers
Logging results to /XXXXXXXX/projects/sparseml/src/sparseml/pytorch/object_detection/runs/train/exp9
Starting training for 5 epochs...
Skipped training due to one-shot: True
One shot checkpoint saved to /XXXXXXXX/projects/sparseml/src/sparseml/pytorch/object_detection/runs/train/exp9/weights/checkpoint-one-shot.pt
```

Note the path to the saved checkpoint,
- [X] export the checkpoint
```bash
sparseml.object_detection.export_onnx \
    --weights PATH/TO/CHECKPOINT
```

- [X] Visualize the checkpoint using Netron

TODO:
- [x] Change base branch to main before merging

**NOTING AGAIN**: [this PR](https://github.com/neuralmagic/yolov5/pull/46) needs to land and and a new wheel needs to be pushed for these changes to work

### CODE REVIEW

**Review by bfineran — COMMENTED** [review]
looks like the max train steps fixes are in here as well do we need a rebase?

Also since a large part of the core train script is touched, please do a quick verification that normal training (ie without one-shot) still runs and update PR description

**Comment by rahul-tuli** [comment]
> looks like the max train steps fixes are in here as well do we need a rebase?
> 
> Also since a large part of the core train script is touched, please do a quick verification that normal training (ie without one-shot) still runs and update PR description

Rebased!

File: src/sparseml/pytorch/object_detection/train.py [hunk 1]
```diff
@@ -427,288 +435,324 @@ def train(hyp, opt, device, callbacks):  # path/to/hyp.yaml or hyp dictionary
...
+                                best_fitness,
+                                fi,
+                            )
+
+            callbacks.run("on_train_end", last, best, plots, epoch, results)
+            LOGGER.info(f"Results saved to {colorstr('bold', save_dir)}")
 
-        callbacks.run("on_train_end", last, best, plots, epoch, results)
-        LOGGER.info(f"Results saved to {colorstr('bold', save_dir)}")
+    else:
```
Comment (bfineran) [inline src/sparseml/pytorch/object_detection/train.py:739]: this is a huge chunk of code to indent under `if not one_shot` and the else happens right before we return.  instead, before the train/eval loop, if one_shot is enabled, can we run the block below and then return?

Comment (rahul-tuli) [reply src/sparseml/pytorch/object_detection/train.py:739]: Thanks! don't know why I didn't do that before,

**Review by bfineran — DISMISSED** [review]
LGTM!

**Review by KSGulin — APPROVED** [review]

**Review by dbogunowicz — APPROVED** [review]

**Pull request merged**